backstory/src/demo/java/com/google/sps/servlets/AnalyzeImageServlet.java

```diff
@@ -0,0 +1,173 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Text;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.protobuf.ByteString;
+import com.google.sps.images.data.AnnotatedImage;
+import com.google.sps.images.ImagesManager;
+import com.google.sps.images.VisionImagesManager;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Prototype Servlet for serving the POST request for the analyze image resource. It uses the Vision
+ * API to analyze the images.
+ */
+@WebServlet("/analyze-image")
+public class AnalyzeImageServlet extends HttpServlet {
+  /**
+   * {@inheritDoc}
+   *
+   * Expecting a post request from Blobstore containing the data fields from the image-upload
+   * form. After having gone through Blobstore, the request will include the image uploaded. The
+   * form in the HTML will connect to the Blobstore URL, which encodes the image and then
+   * redirects the request to this Url. In this doPost, the image is analyzed with Vision API,
+   * and the analysis along with the image URL will be sent to datastore for permanent storage.
+   */
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the input from the form.
+    // Get the URL of the image that the user uploaded to Blobstore.
+    final String imageUrl = getUploadedFileUrl(request, "image-upload-form");
+
+    // Get the raw byte array representing the image from Blobstore
+    final byte[] bytes = getBlobBytes(request, "image-upload-form");
+
+    if (bytes == null || imageUrl == null) {
+      // Redirect back to the HTML page.
+      response.sendError(400, "Please upload a valid image.");
```

Comment:
It would be clearer here to have an early return.
1. Without the early return I had to scoll to the bottom of this method to make sure no other code runs after this.
2. The early return reduces nesting (improves readability) of the block below:

```
if (...) {
  response.sendError(400, "Please upload a valid image.");
  return;
}  // Note: no else block (to reduce nesting).

ImagesManager demoImagesManager = new VisionImagesManager();
...
```

Reply:
I'll make this change.